### PR TITLE
initial ssl_exporter implementation

### LINKED ISSUE #1670: support for ssl_exporter
Hi there 👋🏻  

In an effort to keep it short 😄  I would find it useful for grafana-agent to support [ssl_exporter](https://github.com/ribbybibby/ssl_exporter). Following contributing.md, I'm posting here to see if that's something the maintainers are okay with and if that's the case to claim that contribution. Let me know!

### PR DESCRIPTION
**Pull request by jamesalbert**
#### PR Description

Implements support for [ribbybibby/ssl_exporter](https://github.com/ribbybibby/ssl_exporter)

#### Which issue(s) this PR fixes

#1670 

#### Notes to the Reviewer

I wasn't able to pull down a recent version of ribbybibby/ssl_exporter. I forked it and regenerated the go.mod file which seemed to do the trick. So right now it's all pointed to my repo; I filed an issue over there and intend to keep this PR open for review in the meantime and update when fixed.

linter and tests passed for ssl_exporter on my end, but some tests are failing and I'm not sure if/how I introduced them. Manually tested with configs mentioned in the added docs + some I'm using personally. 

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [ ] Tests updated

### CODE REVIEW

File: docs/user/configuration/integrations/integrations-next/ssl-config.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
++++
+title = "ssl_exporter_config"
++++
+
+# ssl config
+
+The `ssl` block configures the `ssl` integration,
```
Comment (brendamuir) [inline docs/user/configuration/integrations/integrations-next/ssl-config.md:7]:
```suggestion
Use the `ssl` block to configure `ssl` integration.
```

File: docs/user/configuration/integrations/integrations-next/ssl-config.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+which is an embedded version of
```
Comment (brendamuir) [inline docs/user/configuration/integrations/integrations-next/ssl-config.md:8]:
```suggestion
'ssl integration' is a embedded version of
```

File: docs/user/configuration/integrations/integrations-next/ssl-config.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+[`ssl_exporter`](https://github.com/ribbybibby/ssl_exporter). This allows collection of SSL certificate metrics from hosts and files.
```
Comment (brendamuir) [inline docs/user/configuration/integrations/integrations-next/ssl-config.md:9]:
```suggestion
[`ssl_exporter`](https://github.com/ribbybibby/ssl_exporter). This enables the collection of SSL certificate metrics from hosts and files.
```

File: docs/user/configuration/integrations/integrations-next/ssl-config.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+
+## Quick configuration example
+
+To get started, define SSL targets in Grafana agent's integration block:
```
Comment (brendamuir) [inline docs/user/configuration/integrations/integrations-next/ssl-config.md:14]:
```suggestion
Define SSL targets in Grafana agent's integration block:
```

**Review by brendamuir — CHANGES_REQUESTED** [review]

**Comment by tpaschalis** [comment]
Hey James 👋  thanks for taking the time to contribute to Grafana Agent!

I still haven't taken a proper look, but just wanted to mention the test failures are due to the upgrade of `prometheus/common`. Commit https://github.com/prometheus/common/commit/252ff6f235cdd75735af7ddec99ac2f783326bff allowed users to explicitly enable/disable HTTP2, which appears as a new field in our configuration. I think we don't have an issue with it, but if it's an upgrade that came along a new dependency, we should probably update our tests to include this new field.

**Comment by jamesalbert** [comment]
@tpaschalis doh! I thought those upgrades might have been it, I tried reverting some changes, but thanks for confirming 😅 Gonna get some coffee and I'll patch this up

**Comment by jamesalbert** [comment]
amended the tests that were only missing `enable_http2: true`. There still seems to be one more test failing in CI that I can't quite determine the cause. It passes for me locally:

```
go test -timeout 30s -run ^TestOTelConfig$ github.com/grafana/agent/pkg/traces
ok      github.com/grafana/agent/pkg/traces
```

I ran a diff against expected and actual, but it turned up empty aside from the memory addresses which `assert.Equal` shouldn't fail on to my understanding.

**Comment by jamesalbert** [comment]
hm, I'm experiencing different results on subsequent runs without any changes. It fails pretty consistently in ci. I thought by replicating the go test command to match exactly what ci has, it would fail as expected and I could troubleshoot. And it did fail once I ran with the make command above, but then randomly started passing again. 

```
ok      github.com/grafana/agent/pkg/traces     11.269s coverage: 80.8% of statements
```

I think I might be missing something here

**Comment by jamesalbert** [comment]
So I confirmed I wasn't going mad by running the test through a while loop. Out of 10 tries, on the 8th attempt the test failed. All other runs passed. I think what we're running into here is `sort.Slice`

> The sort is not guaranteed to be stable: equal elements may be reversed from their original order. For a stable sort, use SliceStable.

Changed to `sort.SliceStable` hoping that does the trick.

File: docs/user/configuration/integrations/integrations-next/ssl-config.md [hunk 5]
```diff
@@ -5,7 +5,7 @@ title = "ssl_exporter_config"
 # ssl config
 
 The `ssl` block configures the `ssl` integration,
-which is an embedded version of
+'ssl integration' is a embedded version of
```
Comment (brendamuir) [inline docs/user/configuration/integrations/integrations-next/ssl-config.md:8]:
```suggestion
'ssl integration' is an embedded version of
```

**Review by brendamuir — CHANGES_REQUESTED** [review]

**Review by brendamuir — APPROVED** [review]

**Comment by jamesalbert** [comment]
that test function `assertConfigEqual`, while it does seem to do the trick, is really ugly. Hoping folks here who are more skilled in the ways of golang can show me a better way 😅 

but at this point, go.mod's pointing to the right repo now so we should be good on that front

File: docs/user/configuration/integrations/ssl-config.md [hunk 6]
```diff
@@ -97,10 +97,6 @@ Full reference of options:
...
 
-## About SSL modules
+## About ssl_exporter Modules
 
-SSL module is the set of SSL counters to be scraped together from the specific network device.
-
-SSL modules available can be found in the embedded ssl.yml file [here](https://github.com/grafana/agent/blob/main/pkg/integrations/ssl_exporter/common/ssl.yml). If not specified, `if_mib` module is used.
-
-If you need to use custom SSL modules, you can [generate](https://github.com/prometheus/ssl_exporter#generating-configuration) your own ssl.yml file and specify it using `config_file` parameter.
+Read more about the supported modules at [ribbybibby/ssl_exporter](https://github.com/ribbybibby/ssl_exporter#configuration)
```
Comment (brendamuir) [inline docs/user/configuration/integrations/ssl-config.md:102]:
```suggestion
For more information on the supported modules, refer to [ribbybibby/ssl_exporter](https://github.com/ribbybibby/ssl_exporter#configuration)
```

**Review by brendamuir — CHANGES_REQUESTED** [review]

File: docs/user/configuration/integrations/integrations-next/ssl-config.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+# ssl config
+
+Use the `ssl` block to configure `ssl` integration.
+'ssl integration' is an embedded version of
+[`ssl_exporter`](https://github.com/ribbybibby/ssl_exporter). This enables the collection of SSL certificate metrics from hosts and files.
+
+
+## Quick configuration example
+
+Define SSL targets in Grafana agent's integration block:
```
Comment (brendamuir) [inline docs/user/configuration/integrations/integrations-next/ssl-config.md:14]:
```suggestion
Define SSL targets in Grafana Agent's integration block:
```

File: docs/user/configuration/integrations/integrations-next/ssl-config.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  [target: <string>]
+
+  # SSL module to use for polling
+  [module: <string> | default = ""]
+```
+
+
+## About ssl_exporter Modules
+
+Read more about the supported modules at [ribbybibby/ssl_exporter](https://github.com/ribbybibby/ssl_exporter#configuration)
```
Comment (brendamuir) [inline docs/user/configuration/integrations/integrations-next/ssl-config.md:93]:
```suggestion
Fro more information on the supported modules, refer to [ribbybibby/ssl_exporter](https://github.com/ribbybibby/ssl_exporter#configuration)
```

File: docs/user/configuration/integrations/ssl-config.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+# ssl config
+
+The `ssl` block configures the `ssl` integration,
+which is an embedded version of
+[`ssl_exporter`](https://github.com/ribbybibby/ssl_exporter). This allows collection of SSL certificate metrics from hosts and files.
+
+
+## Quick configuration example
+
+To get started, define SSL targets in Grafana agent's integration block:
```
Comment (brendamuir) [inline docs/user/configuration/integrations/ssl-config.md:14]:
```suggestion
Define SSL targets in Grafana Agent's integration block:
```

**Review by brendamuir — CHANGES_REQUESTED** [review]

**Review by brendamuir — APPROVED** [review]

File: pkg/integrations/ssl_exporter/ssl_exporter.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+// SSLTarget represents a target to scrape.
+type SSLTarget struct {
+	Name   string `yaml:"name"`
+	Target string `yaml:"target"`
+	Module string `yaml:"module"`
+}
+
+// Config controls the ssl_exporter integration.
+type Config struct {
+	ConfigFile string      `yaml:"key_file,omitempty"`
```
Comment (tpaschalis) [inline pkg/integrations/ssl_exporter/ssl_exporter.go:31]: In the documentation we're calling this `config_file`; let's settle on one of two names. Also, we might want to mention whether it should always be an absolute path (depending on how the upstream LoadConfig methods uses this field)

File: pkg/traces/config_test.go [hunk 11]
```diff
@@ -1732,10 +1732,46 @@ func sortPipelines(cfg *config.Config) {
...
+	sort.SliceStable(ext, func(i, j int) bool { return ext[i].String() > ext[j].String() })
+	sort.SliceStable(serviceRecv, func(i, j int) bool { return serviceRecv[i].String() > serviceRecv[j].String() })
+	for _, pipeline := range pipelines {
+		sort.SliceStable(pipeline.Exporters, func(i, j int) bool { return pipeline.Exporters[i].String() > pipeline.Exporters[j].String() })
+		sort.SliceStable(pipeline.Receivers, func(i, j int) bool { return pipeline.Receivers[i].String() > pipeline.Receivers[j].String() })
+		sort.SliceStable(pipeline.Processors, func(i, j int) bool { return pipeline.Processors[i].String() > pipeline.Processors[j].String() })
+	}
+}
+
+func assertConfigEqual(t *testing.T, cfg1, cfg2 *config.Config) {
```
Comment (tpaschalis) [inline pkg/traces/config_test.go:1749]: Thanks for fixing the tests that errored with the upgrade of prometheus/common!

Sorry that you had to spend time on this as well. I tried to see whether this fixes the flakiness once and for all; it seems it's a lot better but still doesn't 100% pass. 
```
$ go test -run=TestOTelConfig -count=100 | grep "FAIL" | wc -l
      38
```

My 2c is to leave this another PR, keep track of it on #1654 and only focus on the SSL exporter. WDYT? I'll get the word out to people more experienced with the Tracing side of things; I have a hunch that there's a deeper issue with how we use pointers as map keys in the OTEL configuration but they can let us know.

Comment (jamesalbert) [reply pkg/traces/config_test.go:1749]: hey it's been a fun problem to (try to) solve 😎 

That sounds good to me @tpaschalis, should I keep my changes thus far or should I revert my changes to config_test.go?

Comment (tpaschalis) [reply pkg/traces/config_test.go:1749]: IMHO let's revert changes to `pkg/traces/config_test.go` and focus on the exporter itself.

File: pkg/integrations/ssl_exporter/ssl_exporter.go (hunk 10 above)
Comment (jamesalbert) [reply pkg/integrations/ssl_exporter/ssl_exporter.go]: not sure why that was key_file, changed it to config_file

**Comment by jamesalbert** [comment]
> I'm playing around with the exporter, so more detailed comments are coming along, but a high-level note; I think we might be able to drop the v2 and use a ["shim"](https://github.com/grafana/agent/blob/8d090877f2bd6bb47052ee420cc2e8720b25dde6/pkg/integrations/v2/metricsutils/versionshim.go#L20). Here's how we do it for other [exporters](https://github.com/grafana/agent/blob/8d090877f2bd6bb47052ee420cc2e8720b25dde6/pkg/integrations/redis_exporter/redis_exporter.go#L121) for example and I think it will simplify things. WDYT?

yo @tpaschalis, I actually took a stab at this last night and today. ssl_exporter doesn't have an exporter the way other integrations do so it was really tricky to get this going. The old way, v1 worked fine, but v2 didn't do anything. So I tried my best making a mini exporter that collects everything from the registry and sends it to the channel that way. The tests for ssl_exporter itself pass on my end and I've tested this minimally against some local resources, so I'm just checking this in right now before I go too far down a rabbit hole 🙃

**Comment by tpaschalis** [comment]
Hey man, sorry for the intermittent replies, most of the team was out of office these past two weeks.

I looked through your second approach, and you're right, the shim doesn't play well with multi-target exporters such as this one as it doesn't propagate the query params.

I'm cc-ing @mattdurham for his opinion here on what course to follow before committing to something :) I think v1 is on a good path, just making sure which way to go for v2.

**Comment by mattdurham** [comment]
Apologies, was at the offsite last week and spent the beginning of the week battling a thankfully just a cold. I have this on my radar to take a look at today.

**Comment by jamesalbert** [comment]
Hey no worries at all. Hope y'all had fun at the offsite and that you're feeling better Matt! Let me know where you want to go with this and I'm happy to add/remove where necessary.

**Comment by mattdurham** [comment]
@jamesalbert should have a branch with some changes for V2 tomorrow for you to take a look at

**Comment by mattdurham** [comment]
By tomorrow I mean today, ended up making some significant changes. Porting the changes into a v2 integration now

**Comment by jamesalbert** [comment]
yo @mattdurham all that sounds good. How do y'all want to continue with this? Should I merge in that branch into my fork's main? Or do you want to hold off until further cleanup? 

The first thing that stands out to me though is ssl.go in both v1 and v2. Is there anyway we can share between the two? I haven't gotten a chance to fully dig in yet so I'm not sure if you adjusted anything in v2's ssl.go, it looks like an added file to me at this point, but I'll try getting some time in here soon

**Comment by mattdurham** [comment]
There are differences due to how v2/v1 work, that being said we can probably simplify it. I can take a spin at it tomorrow morning and see what I can simplify on my end.

**Comment by jamesalbert** [comment]
haven't abandoned this :) other things popped up. Will revisit soon

**Pull request closed without merging**